### PR TITLE
Fix parse_url() invalid behavior with UTF-8

### LINKED ISSUE #507: Diacritics from stream with LocalAdapter
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Gaufrette version  | 0.3.1
| PHP version  | 7.1.5 and 7.1.6
| System  | OSX Sierra and Ubuntu 16.04

I recently encountered a problem with diacritics when trying to access a file from a stream.

I setup `uploads` as a filesystem with a local adapter.

the following is working:
```php
file_get_contents("gaufrette://uploads/foo.txt");
```

the following is not working:
```php
file_get_contents("gaufrette://uploads/baré.txt");
```

NB: files exist on filesystem and sources files are encoded in UTF-8 by default.

### PR DESCRIPTION
**Pull request by phcorp**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | this fixes #507
| License       | MIT

### CODE REVIEW

**Comment by NiR-** [comment]
Could you write a test for `Path::parseUrl` please?

File: src/Gaufrette/Util/Path.php [hunk 1]
```diff
@@ -86,4 +86,39 @@ public static function dirname($path)
...
+     *
+     * @see https://secure.php.net/manual/fr/function.parse-url.php#114817
+     *
+     * @param string $url to parse
+     *
+     * @return array
+     */
+    public static function parseUrl($url)
+    {
+        $enc_url = preg_replace_callback(
```
Comment (NiR-) [inline src/Gaufrette/Util/Path.php]: Could you use a better name? Also, you should  use CamelCase.

**Comment by phcorp** [comment]
I added a unit and a functional test, hope it suits.

**Comment by NiR-** [comment]
Sorry to postpone your PR a bit, even after requesting changes but I would like to see opinions of other contributors/users for this one.

What troubles me: at first I thought it was a bug (an unsolved php bug, fixed by userland code) but after double-checking I now think it is a new feature. Indeed `parse_url` does not support non-ascii characters but that because it strictly implements [RFC 3986](http://www.faqs.org/rfcs/rfc3986.html). This RFC is cited in the |`parse_url` doc](https://php.net/parse_url) and moreover in [this related bug report](https://bugs.php.net/bug.php?id=52923).

File: src/Gaufrette/Util/Path.php [hunk 2]
```diff
@@ -86,4 +86,39 @@ public static function dirname($path)
...
     }
+
+    /**
+     * UTF-8 aware parse_url() replacement.
+     *
+     * @param string $url to parse
+     *
+     * @return bool|array
+     *
+     * @see https://secure.php.net/manual/fr/function.parse-url.php#114817
```
Comment (nicolasmure) [inline src/Gaufrette/Util/Path.php]: Can you change this link to use the english locale please ? (`manual/fr` => `manual/en`) : https://secure.php.net/manual/en/function.parse-url.php#114817

File: src/Gaufrette/Util/Path.php [hunk 3]
```diff
@@ -86,4 +86,39 @@ public static function dirname($path)
...
+        );
+
+        $parts = parse_url($encodedUrl);
+
+        if($parts === false)
+        {
+            return false;
+        }
+
+        foreach($parts as $name => $value)
```
Comment (nicolasmure) [inline src/Gaufrette/Util/Path.php]: PSRs : `foreach ($parts as $name) {`

File: src/Gaufrette/Util/Path.php [hunk 4]
```diff
@@ -86,4 +86,39 @@ public static function dirname($path)
...
+            function ($matches)
+            {
+                return urlencode($matches[0]);
+            },
+            $url
+        );
+
+        $parts = parse_url($encodedUrl);
+
+        if($parts === false)
```
Comment (nicolasmure) [inline src/Gaufrette/Util/Path.php]: Please follow the PSRs : `if (false === $parts) {` ;)

File: tests/Gaufrette/Unit/Util/PathTest.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            'host' => 'föô',
+            'path' => '/bàr',
+            'query' => 'bäz=1',
+            'fragment' => 'qùx',
+        ];
+        $parts = Path::parseUrl($url);
+
+        $this->assertEquals($expected, $parts);
+    }
+    /**
```
Comment (nicolasmure) [inline tests/Gaufrette/Unit/Util/PathTest.php]: Please insert one empty line between the 2 functions.

**Comment by nicolasmure** [comment]
Hello @phcorp ,

We thank you for your contribution, but after a discussion and [as explained](https://github.com/KnpLabs/Gaufrette/pull/508#issuecomment-311176207) by @NiR- , we finally won't merge this PR as UTF-8 chars are not supported by the [RFC 3986](http://www.faqs.org/rfcs/rfc3986.html).

We recommend you to store files named with letters from ASCII table.

Thanks.

**Pull request closed without merging**